tests/lib/rule-tester/rule-tester.js

```diff
@@ -463,6 +463,92 @@ describe("RuleTester", () => {
+                    valid: [{
+                        code: "const onlyValid = 42;",
+                        [hookName]: hookForValid
+                    }],
+                    invalid: [{
+                        code: "var onlyValid = 42;",
+                        errors: [/^Bad var/u],
+                        output: " onlyValid = 42;",
+                        [hookName]: hookForInvalid
+                    }]
+                });
+                sinon.assert.calledOnce(hookForValid);
+                sinon.assert.calledOnce(hookForInvalid);
+            });
+
+            it(`${hookName} should cause test to fail when it throws error`, () => {
+                const hook = sinon.stub().throws(new Error("Something happended"));
+
+                ruleTester = new RuleTester();
+                assert.throws(() => ruleTester.run(ruleName, rule, {
+                    valid: [{
+                        code: "const onlyValid = 42;",
+                        [hookName]: hook
+                    }],
+                    invalid: []
+                }), "Something happended");
+            });
+
+            it(`${hookName} should throw when not a function is assigned`, () => {
+                ruleTester = new RuleTester();
+                assert.throws(() => ruleTester.run(ruleName, rule, {
+                    valid: [{
+                        code: "const onlyValid = 42;",
+                        [hookName]: 42
+                    }],
+                    invalid: []
+                }), `Optional test case property '${hookName}' must be a function`);
+            });
+        });
+
+        it("should call both before() and after() hooks even when the case failed", () => {
+            const hookBefore = sinon.stub();
+            const hookAfter = sinon.stub();
+
+            ruleTester = new RuleTester();
+            assert.throws(() => ruleTester.run(ruleName, rule, {
+                valid: [{
+                    code: "var onlyValid = 42;",
+                    before: hookBefore,
+                    after: hookAfter
+                }],
+                invalid: []
+            }));
+            sinon.assert.calledOnce(hookBefore);
+            sinon.assert.calledOnce(hookAfter);
+        });
+
+        it("should call after() hook even when before() throws", () => {
+            const hookBefore = sinon.stub().throws(new Error("Something happended in before()"));
+            const hookAfter = sinon.stub();
+
+            ruleTester = new RuleTester();
+            assert.throws(() => ruleTester.run(ruleName, rule, {
+                valid: [{
+                    code: "const onlyValid = 42;",
+                    before: hookBefore,
+                    after: hookAfter
+                }],
+                invalid: []
+            }), "Something happended in before()");
+            sinon.assert.calledOnce(hookBefore);
+            sinon.assert.calledOnce(hookAfter);
+        });
+    });
+
```

Comment:
To cover ourselves, can you please add some tests that:

1. Validate `after` is called in invalid cases
2. Validate `after` is called when there's a syntax error in `code`.

Reply:
Those tests I actually took care to implement, @nzakas 

## For the first one — `after` is called in invalid cases

See that all hooks are covered: both before and after:
https://github.com/eslint/eslint/blob/aa04ab859fa18d8a0c4ff2d49c547de8e6467206/tests/lib/rule-tester/rule-tester.js#L470

Notice that the hook is assigned to invalid case:
https://github.com/eslint/eslint/blob/aa04ab859fa18d8a0c4ff2d49c547de8e6467206/tests/lib/rule-tester/rule-tester.js#L481-L485

And finally there is an assertion that it's called:
https://github.com/eslint/eslint/blob/aa04ab859fa18d8a0c4ff2d49c547de8e6467206/tests/lib/rule-tester/rule-tester.js#L489

## For the second — `after` is called for an error in `code`

Notice again that both hooks are tested:
https://github.com/eslint/eslint/blob/aa04ab859fa18d8a0c4ff2d49c547de8e6467206/tests/lib/rule-tester/rule-tester.js#L470

Then notice this test, where `code` does not pass the `no-var` rule and throws, but call of `after` is asserted:
https://github.com/eslint/eslint/blob/aa04ab859fa18d8a0c4ff2d49c547de8e6467206/tests/lib/rule-tester/rule-tester.js#L517-L524

-----------

In case I didn't understand what you mean, or you meant something else, please clarify and I will make adjustments.

Reply:
@RobinTail 

1. The first test covers both `valid` and `invalid` items, but the other four tests cover only `valid` items. It would be nice to have all test cases for both `valid` and `invalid` items since they're handled in different code paths.
2. Syntax error in `code` means that `code` is not parsable, i.e., the test fails due to a parsing error (e.g., `code: "foo bar"`). The test where `code` does not pass the `no-var` rule is good, this would be an additional one where `code` is something that isn't valid JS.

Reply:
Thank you for the explanation, @mdjermanovic .
I'll do my best to improve it.

Reply:
Addressed in 7543787 and 7de42961e, @mdjermanovic  && @nzakas